crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -436,50 +468,95 @@ impl Dispatch<wl_keyboard::WlKeyboard, ()> for WaylandClientState {
                     keymap_state.mod_name_is_active(xkb::MOD_NAME_LOGO, xkb::STATE_MODS_EFFECTIVE);
+
+                state.modifiers.shift = shift;
+                state.modifiers.alt = alt;
+                state.modifiers.control = control;
+                state.modifiers.command = command;
             }
             wl_keyboard::Event::Key {
                 key,
                 state: WEnum::Value(key_state),
                 ..
             } => {
+                let focused_window = &state.keyboard_focused_window;
+                let Some(focused_window) = focused_window else {
+                    return;
+                };
+
                 let keymap_state = state.keymap_state.as_ref().unwrap();
                 let keycode = Keycode::from(key + MIN_KEYCODE);
-
-                let focused_window = &state.keyboard_focused_window;
-                if let Some(focused_window) = focused_window {
-                    match key_state {
-                        wl_keyboard::KeyState::Pressed => {
-                            focused_window.handle_input(PlatformInput::KeyDown(KeyDownEvent {
-                                keystroke: Keystroke::from_xkb(
-                                    keymap_state,
-                                    state.modifiers,
-                                    keycode,
-                                ),
-                                is_held: false, // todo!(linux)
-                            }));
+                let keysym = keymap_state.key_get_one_sym(keycode);
+
+                match key_state {
+                    wl_keyboard::KeyState::Pressed => {
+                        let input = PlatformInput::KeyDown(KeyDownEvent {
+                            keystroke: Keystroke::from_xkb(keymap_state, state.modifiers, keycode),
+                            is_held: false, // todo!(linux)
+                        });
+
+                        focused_window.handle_input(input.clone());
+
+                        if !keysym.is_modifier_key() {
+                            state.repeat.current_id += 1;
+                            state.repeat.current_keysym = Some(keysym);
+
+                            let rate = state.repeat.rate;
+                            let delay = state.repeat.delay;
+                            let id = state.repeat.current_id;
+                            let keysym = state.repeat.current_keysym;
+                            let state_container = state_container.clone();
+
+                            state
+                                .platform_inner
+                                .foreground_executor
+                                .spawn(async move {
+                                    let mut wait_time = Duration::from_millis(delay as u64);
+
+                                    loop {
+                                        Timer::after(wait_time).await;
+
+                                        let state = state_container.0.lock();
```

Comment:
I think it would be good to add a comment here with a disclamer that calling await after this line could result in dead locks.

Reply:
Nevermind, turns out clippy would be smart enough to catch that anyways.